修复本地存在代理情况下,webui因网络问题启动失败

taoharry opened the pull request:
# What does this PR do?

Fixes #6739 (issue)
修复本地存在代理情况下,webui因网络问题启动失败
 
(llama-factory) % llamafactory-cli webui 
* Running on local URL:  http://0.0.0.0:7860
Traceback (most recent call last):
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_transports/default.py", line 101, in map_httpcore_exceptions
    yield
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_transports/default.py", line 250, in handle_request
    resp = self._pool.handle_request(req)
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py", line 256, in handle_request
    raise exc from None
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py", line 236, in handle_request
    response = connection.handle_request(
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpcore/_sync/http_proxy.py", line 206, in handle_request
    return self._connection.handle_request(proxy_request)
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpcore/_sync/connection.py", line 103, in handle_request
    return self._connection.handle_request(request)
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 136, in handle_request
    raise exc
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 106, in handle_request
    ) = self._receive_response_headers(**kwargs)
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 177, in _receive_response_headers
    event = self._receive_event(timeout=timeout)
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 231, in _receive_event
    raise RemoteProtocolError(msg)
httpcore.RemoteProtocolError: Server disconnected without sending a response.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/anaconda3/envs/llama-factory/bin/llamafactory-cli", line 8, in <module>
    sys.exit(main())
  File "/Users/mac/rag_ai/LLaMA-Factory/src/llamafactory/cli.py", line 122, in main
    run_web_ui()
  File "/Users/mac/rag_ai/LLaMA-Factory/src/llamafactory/webui/interface.py", line 94, in run_web_ui
    create_ui().queue().launch(share=gradio_share, server_name=server_name, inbrowser=True)
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/gradio/blocks.py", line 2645, in launch
    resp = httpx.get(
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_api.py", line 195, in get
    return request(
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_api.py", line 109, in request
    return client.request(
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_client.py", line 825, in request
    return self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_client.py", line 914, in send
    response = self._send_handling_auth(
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_client.py", line 942, in _send_handling_auth
    response = self._send_handling_redirects(
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_client.py", line 979, in _send_handling_redirects
    response = self._send_single_request(request)
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_client.py", line 1014, in _send_single_request
    response = transport.handle_request(request)
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_transports/default.py", line 249, in handle_request
    with map_httpcore_exceptions():
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/opt/anaconda3/envs/llama-factory/lib/python3.10/site-packages/httpx/_transports/default.py", line 118, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.RemoteProtocolError: Server disconnected without sending a response.
## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/hiyouga/LLaMA-Factory/blob/main/.github/CONTRIBUTING.md)?
- [x] Did you write any new necessary tests?

hiyouga commented on src/llamafactory/webui/interface.py:
```diff
@@ -88,10 +90,26 @@ def create_web_demo() -> "gr.Blocks":
     gradio_share = is_env_enabled("GRADIO_SHARE")
     server_name = os.getenv("GRADIO_SERVER_NAME", "[::]" if gradio_ipv6 else "0.0.0.0")
-    create_ui().queue().launch(share=gradio_share, server_name=server_name, inbrowser=True)
+    httpx.HTTPCORE_OPTS = {"trust_env": False}
```
Could you explain the reason of changing the following lines?

hiyouga commented on src/llamafactory/webui/interface.py:
```diff
@@ -88,10 +90,26 @@ def create_web_demo() -> "gr.Blocks":
 
 
 def run_web_ui() -> None:
+    os.environ["no_proxy"] = "localhost,127.0.0.1,0.0.0.0"
```
how about remove http_proxy and https_proxy (for ipv6 case)

taoharry commented:
When enabling global proxy (e.g., VPN) in the local environment, LLaMA-Factory's Web UI fails to start properly. Debugging revealed the issue stemmed from proxy settings in the httpx library: when global system proxies are active, httpx defaults to using them, mistakenly routing local requests (e.g., to localhost) through the proxy server.


Here’s a summary of the changes made to resolve proxy-related issues, with technical explanations:
1. Root Cause


System-wide proxy settings interfered with local server communication for LLaMA-Factory’s Web UI, leading to connection timeouts and disconnections.


Explanation: Global proxies (e.g., VPN) redirect all network traffic, including local requests to localhost/127.0.0.1. The httpx library, which LLaMA-Factory uses for HTTP calls, implicitly inherited these settings, causing local requests to fail (proxy servers cannot route traffic to the local machine).
2. Modified Files & Changes
interface.py


Added no_proxy environment variable to exclude local addresses (localhost, 127.0.0.1, 0.0.0.0).
Explanation: no_proxy explicitly tells the system, "Bypass proxies for these local IPs". This ensures requests to the local server (where LLaMA-Factory runs) are never routed through the proxy, fixing direct communication.
Configured httpx client with trust_env=False.
Explanation: By default, httpx "trusts" system environment variables (e.g., HTTP_PROXY). Setting trust_env=False forces httpx to ignore global proxy settings entirely, ensuring only intentional proxies (if configured) are used—critical for local development where proxies should not affect localhost traffic.
Optimized Gradio server startup parameters (server_name).
Explanation: Explicitly defining server_name="0.0.0.0"  ensures the Gradio web server binds to all interfaces correctly. This prevents  misconfigured listening addresses that could exacerbate proxy-related errors.
3. Impact


These changes force local traffic to bypass global proxies, allowing the Web UI to communicate directly with the local LLaMA-Factory server.


Result: Eliminated connection timeouts, stable Web UI startup, and consistent local server access—even when global proxies are active.


Technical Note:
This fix targets a common issue in Python web apps: httpx/requests inheriting unintended system proxies for localhost. The no_proxy + trust_env=False combo is a standard workaround for local development environments with enforced global proxy settings (e.g., corporate networks, VPNs).




---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2025年03月17日 15:07 |
| 收件人 | ***@***.***> |
| 抄送人 | ***@***.***> ,
***@***.***> |
| 主题 | Re: [hiyouga/LLaMA-Factory] 修复本地存在代理情况下,webui因网络问题启动失败 (PR #7332) |

@hiyouga commented on this pull request.

In src/llamafactory/webui/interface.py:

>      gradio_ipv6 = is_env_enabled("GRADIO_IPV6")
     gradio_share = is_env_enabled("GRADIO_SHARE")
     server_name = os.getenv("GRADIO_SERVER_NAME", "[::]" if gradio_ipv6 else "0.0.0.0")
-    create_ui().queue().launch(share=gradio_share, server_name=server_name, inbrowser=True)
+    httpx.HTTPCORE_OPTS = {"trust_env": False}


Could you explain the reason of changing the following lines?

In src/llamafactory/webui/interface.py:

> @@ -88,10 +90,26 @@ def create_web_demo() -> "gr.Blocks":
 
 
 def run_web_ui() -> None:
+    os.environ["no_proxy"] = "localhost,127.0.0.1,0.0.0.0"


how about remove http_proxy and https_proxy (for ipv6 case)

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***>
